File: _B_pendulum/simulink_old/hw13/pendulumParamHW13.m
% inverted Pendulum parameter file
addpath ./.. % adds the parent directory to the path
pendulumParam % general pendulum parameters

% saturation limits
P.F_max = 5;                   % Max Force, N

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% observer based control with integrator
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% tuning parameters for control
tr_z = 1.5;            % rise time for position 
tr_theta = 0.5;        % rise time for angle
zeta_z   = 0.707;      % damping ratio position
zeta_th  = 0.707;      % damping ratio angle
integrator_pole = -3; % integrator pole


% state space equations
P.A = [...
    0, 0, 1, 0;...
    0, 0, 0, 1;...
    0, -3*P.m1*P.g/4/(.25*P.m1+P.m2), -P.b/(.25*P.m1+P.m2), 0;...
    0, 3*(P.m1+P.m2)*P.g/2/(.25*P.m1+P.m2)/P.ell, ...
                              3*P.b/2/(.25*P.m1+P.m2)/P.ell, 0;...
];
P.B = [0; 0; 1/(.25*P.m1+P.m2); -3/2/(.25*P.m1+P.m2)/P.ell ];
P.C = [...
    1, 0, 0, 0;...
    0, 1, 0, 0;...
    ];
% form augmented system
Cout = [1,0,0,0];
A1 = [P.A, zeros(4,1); -Cout, 0];
B1 = [P.B; 0];

% compute control gains
wn_th = 2.2/tr_theta; % natural frequency - angle
wn_z = 2.2/tr_z; % natural frequency - position

% tuning parameters for observer
wn_th_obs   = 5*wn_th;
wn_z_obs    = 5*wn_z;

des_char_poly = conv(...
    conv([1,2*zeta_z*wn_z,wn_z^2],...
         [1,2*zeta_th*wn_th,wn_th^2]),...
    poly(integrator_pole));
des_poles = roots(des_char_poly);

% is the system controllable?
if rank(ctrb(A1,B1))~=5
    disp('System Not Controllable'); 
else % if so, compute gains
    K1   = place(A1,B1,des_poles); 
    P.K  = K1(1:4);
    P.ki = K1(5);
end

% computer observer gains
des_obsv_char_poly = conv(...
    [1,2*zeta_z*wn_z_obs,wn_z_obs^2],...
    [1,2*zeta_th*wn_th_obs,wn_th_obs^2]);
des_obsv_poles = roots(des_obsv_char_poly);

% is the system observable?
if rank(obsv(P.A,P.C))~=4
    disp('System Not Observable'); 
else % if so, compute gains
    P.L = place(P.A', P.C', des_obsv_poles)';
end

sprintf('K: [%f, %f, %f, %f]\nki: %f\nL: [%f, %f, %f, %f]^T',...
    P.K(1), P.K(2), P.K(3), P.K(4), P.ki,...
        P.L(1), P.L(2), P.L(3), P.L(4))

File: _B_pendulum/python/hw13/pendulumParamHW13.py
# Inverted Pendulum Parameter File
import pendulumParam as P
import numpy as np
from scipy import signal
import control as cnt

# sample rate of the controller
Ts = P.Ts

# saturation limits
F_max = 5.0                # Max Force, N

####################################################
#                 State Space
####################################################
# tuning parameters
tr_z = 1.5                   # rise time for position
tr_theta = 0.5               # rise time for angle
zeta_z = 0.707             # damping ratio position
zeta_th = 0.707             # damping ratio angle
integrator_pole = -2.0       # integrator pole
tr_z_obs = tr_z/5.0          # rise time for observer - position
tr_theta_obs = tr_theta/5.0  # rise time for observer - angle

# State Space Equations
# xdot = A*x + B*u
# y = C*x

A = np.matrix([[0.0, 0.0, 1.0, 0.0],
               [0.0, 0.0, 0.0, 1.0],
               [0.0, -3*P.m1*P.g/4/(.25*P.m1+P.m2),
                -P.b/(.25*P.m1+P.m2), 0.0],
               [0.0, 3*(P.m1+P.m2)*P.g/2/(.25*P.m1+P.m2)/P.ell,
                3*P.b/2/(.25*P.m1+P.m2)/P.ell, 0.0]])

B = np.matrix([[0.0],
               [0.0],
               [1/(.25*P.m1+P.m2)],
               [-3.0/2/(.25*P.m1+P.m2)/P.ell]])

C = np.matrix([[1.0, 0.0, 0.0, 0.0],
               [0.0, 1.0, 0.0, 0.0]])

# form augmented system
Cr = C[0, :]

# Augmented Matrices
A1 = np.concatenate((
        np.concatenate((A, np.zeros((4, 1))), axis=1),
        np.concatenate((-Cr, np.matrix([[0.0]])), axis=1)),
        axis=0)
B1 = np.concatenate((B, np.matrix([[0.0]])), axis=0)

# computer control gains
wn_th = 2.2/tr_theta  # natural frequency for angle
wn_z = 2.2/tr_z  # natural frequency for position
des_char_poly = np.convolve(
    np.convolve([1, 2*zeta_z*wn_z, wn_z**2],
                [1, 2*zeta_th*wn_th, wn_th**2]),
    np.poly([integrator_pole]))
des_poles = np.roots(des_char_poly)

# Compute the gains if the system is controllable
if np.linalg.matrix_rank(cnt.ctrb(A1, B1)) != 5:
    print("The system is not controllable")
else:
    K1 = cnt.acker(A1, B1, des_poles)
    K = K1[0, 0:4].reshape(1,4)
    ki = K1[0, 4].reshape(1,1)

# computer observer gains
wn_z_obs = 2.2/tr_z_obs
wn_th_obs = 2.2/tr_theta_obs
des_obs_char_poly = np.convolve(
    [1, 2*zeta_z*wn_z_obs, wn_z_obs**2],
    [1, 2*zeta_th*wn_th_obs, wn_th_obs**2])
des_obs_poles = np.roots(des_obs_char_poly)

# Compute the gains if the system is observable
if np.linalg.matrix_rank(cnt.ctrb(A.T, C.T)) != 4:
    print("The system is not observable")
else:
    L = signal.place_poles(A.T, C.T, des_obs_poles).gain_matrix.T

print('K: ', K)
print('ki: ', ki)
print('L^T: ', L.T)
